File: src/core/paths.ts
import { homedir } from 'node:os';
import path from 'node:path';

/**
 * homer 本地工作区的全部路径。
 * 布局（DESIGN §2.1）：
 *   <home>/homer.json   主配置（随 git 仓库走）
 *   <home>/store/       adapter 分类快照
 *   <home>/state.json   本地同步状态（不入库）
 *   <home>/backups/     应用远端变更前的备份（不入库，见 docs/m2-plan.md §2.0-2）
 *   <home>/secrets/     age 密文 vault（**入库**，一密钥一 <name>.age，见 docs/m3-plan.md §2.0-2 / D3）
 *   <home>/keys/        本机 age 私钥（`age.txt`，0600；**不入库**，见 docs/m3-plan.md §2.0-2 / D2）
 */
export interface HomerPaths {
  home: string;
  storeDir: string;
  configFile: string;
  stateFile: string;
  backupsDir: string;
  secretsDir: string;
  keysDir: string;
}

/**
 * 展开开头的 `~` / `~/`。只处理开头，不做全局替换（与 shell 语义一致）。
 * 导出给 adapter 侧复用（scan.ts 曾自带一份同名私有实现，属重复代码）。
 */
export function expandHome(input: string): string {
  if (input === '~') return homedir();
  if (input.startsWith('~/') || input.startsWith('~\\')) {
    return path.join(homedir(), input.slice(2));
  }
  return input;
}

/**
 * 解析 homer 工作区路径。
 *
 * 优先级：显式传入的 `env.HOMER_HOME` > `process.env.HOMER_HOME` > `~/.homer`。
 * `HOMER_HOME` 覆盖是测试不碰真实 `~/.homer` 的关键；空字符串视为未设置。
 * 返回值中的 `home` 一律是绝对路径（相对路径按 cwd 解析）。
 */
export function getHomerPaths(env: { HOMER_HOME?: string | undefined } = process.env): HomerPaths {
  const raw = env.HOMER_HOME;
  const home =
    typeof raw === 'string' && raw.trim() !== ''
      ? path.resolve(expandHome(raw.trim()))
      : path.join(homedir(), '.homer');

  return {
    home,
    storeDir: path.join(home, 'store'),
    configFile: path.join(home, 'homer.json'),
    stateFile: path.join(home, 'state.json'),
    backupsDir: path.join(home, 'backups'),
    secretsDir: path.join(home, 'secrets'),
    keysDir: path.join(home, 'keys'),
  };
}

File: internal/core/paths.go
package core

import (
	"os"
	"path/filepath"
	"strings"
)

// HomerPaths contains every path in a Homer workspace.
type HomerPaths struct {
	Home       string
	StoreDir   string
	ConfigFile string
	StateFile  string
	BackupsDir string
	SecretsDir string
	KeysDir    string
}

// expandHome expands only a leading ~ or ~/ (and the Windows spelling ~\\).
// Other occurrences of ~ are ordinary path characters, matching shell
// expansion semantics and the TypeScript implementation.
func expandHome(input string) string {
	home, err := os.UserHomeDir()
	if err != nil {
		home = ""
	}
	if input == "~" {
		return home
	}
	if strings.HasPrefix(input, "~/") || strings.HasPrefix(input, `~\`) {
		return filepath.Join(home, input[2:])
	}
	return input
}

// ExpandHome is the exported form used by adapter and command packages.
func ExpandHome(input string) string { return expandHome(input) }

// GetHomerPaths resolves the Homer workspace. A non-empty, trimmed
// HOMER_HOME returned by env wins over ~/.homer; relative values are resolved
// against the current working directory. Passing nil uses os.Getenv.
func GetHomerPaths(env func(string) string) HomerPaths {
	if env == nil {
		env = os.Getenv
	}

	raw := strings.TrimSpace(env("HOMER_HOME"))
	home := ""
	if raw != "" {
		home = expandHome(raw)
		if absolute, err := filepath.Abs(home); err == nil {
			home = absolute
		}
	} else {
		userHome, err := os.UserHomeDir()
		if err != nil {
			userHome = "."
		}
		home = filepath.Join(userHome, ".homer")
	}

	return HomerPaths{
		Home:       home,
		StoreDir:   filepath.Join(home, "store"),
		ConfigFile: filepath.Join(home, "homer.json"),
		StateFile:  filepath.Join(home, "state.json"),
		BackupsDir: filepath.Join(home, "backups"),
		SecretsDir: filepath.Join(home, "secrets"),
		KeysDir:    filepath.Join(home, "keys"),
	}
}
